src/glossary.adoc

```diff
@@ -0,0 +1,116 @@
+The number of elements accepted by a single argument of a macro, or produced by a macro. The
+arguments and results of a macro are streams, and each can be constrained to have exactly-one
+(`*!*`), zero-or-one (`*?*`), zero-or-more (`*{asterisk}*`, `*\...*`), or one-or-more (`*+*`)
+elements.
+
+declaration::
+The association of a name with an entity (for example, a module or macro). See also _definition_. Not
+all declarations are definitions: some introduce new names for existing entities.
+
+definition::
+The specification of a new entity. Most (but not all) definitions are declarations.
+
+directive::
+A unit of data in an Ion document that affects the encoding context, and thus the way the
+document’s data is decoded. In Ion 1.0 there are two directives: the Ion Version Marker, and
+top-level `$ion_symbol_table` structs. Ion 1.1 adds the `$ion_encoding` directive.
+
+document::
+A stream of octets conforming to either the Ion text or binary specification. Can consist of
+multiple _segments_, perhaps using varying versions of the Ion specification. A document doesn’t
+necessarily exist as a file, and isn’t necessarily finite.
+
+E-expression::
+See _encoding expression_.
+
+encoding directive::
+In Ion 1.1, a top-level `$ion_encoding` struct that controls the segment immediately following it.
+The 1.0 `$ion_symbol_table` directive is effectively a less capable alternative syntax.
+
+encoding environment::
+The context-specific data maintained by an Ion implementation while encoding or decoding data. In
+Ion 1.0 this consists of the local symbol table; in Ion 1.1 it also has the local module table
+and a collection of available modules.
+
+encoding expression::
+The invocation of a macro in encoded data, aka E-expression. Starts with a macro reference
+denoting the function to invoke. The Ion text format uses “smile syntax” `(:__macro__ …)` to denote
+E-expressions. Ion binary devotes a large number of opcodes to E-expressions, so they can be
+compact.
+
+macro::
+A transformation function that accepts some number of streams of values, and produces a stream of
+values.
+
+macro definition::
+Specifies a macro in terms of a _signature_ and a _template_.
+
+macro reference::
+Identifies a macro for invocation, renaming, or exporting. Must always be unambiguous. Lexically
+scoped, and never a “forward reference” to a macro that’s declared later in the document.
+
+module::
+The data entity that defines and exports both symbols and macros. Modules are imported by
+encoding directives then installed into the local symbol and/or macro tables.
+
+parameter::
+A named input to a macro, as defined by its signature. At expansion time a parameter produces a
+stream of values.
+
+qualified macro reference::
+A macro reference that consists of a module name and either a macro name exported by that module,
+or a numeric address within the range of the module’s exported macro table. In text, these look
+like `:__module-name__:__name-or-address__`.
```

Comment:
The *`:name-or-address`* form is also a qualified reference, yes?  Is this maybe a _local qualified macro reference_?

Reply:
I'm not sure that "local" works well, since `:name` could refer to any module, including `$ion`.

Reply:
True, but isn't it unambiguous when `:name` refers to a macro versus a module (I cannot invoke a module)?  What do we call that sort of identifier?

For *macro references*, I see three kinds:

* _unqualified macro references_ - these are denoted without a `:` prefix.  E.g., `foobar`.
* _qualified macro references_ - these are denoted with _module name_ and prefixed/delimited with `:`.  E.g., `:something:foobar`
* _partially qualified macro references_ (not sold on that term) - these are denoted with the `:` prefix and no module name.  Such references are resolved against the _current macro table_.  E.g., `:foobar`